include/deal.II/numerics/kdtree_distance.h

```diff
@@ -0,0 +1,266 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef _dealii__numerics_kdtree_distance_h
+#define _dealii__numerics_kdtree_distance_h
+
+#include <deal.II/base/config.h>
+
+#  ifdef DEAL_II_WITH_NANOFLANN
+
+#include <deal.II/base/point.h>
+
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
+#include <nanoflann.hpp>
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
+
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A wrapper for the nanoflann library, used to compute the distance from a
+ * collection of points, and to efficiently return nearest neighbors to a
+ * target point. This function uses nanoflann to efficiently partition the
+ * space in a tree. The cost of each query is then roughly of order log(n),
+ * where n is the number of points stored in this class.
+ *
+ * The wrapper provides methods that give access to some of thefunctionalities
+ * of the nanoflann library, like searching the n nearest neighbors, or
+ * searching the points that fall within a raidius of a target point.
+ */
+template<int dim>
+class KDTreeDistance
```

Comment:
https://en.wikipedia.org/wiki/K-d_tree

I'll add a comment to the documentation.